File: python/quickstart/src/agnt5_quickstart/workflows.py
"""The `digest` workflow — fan out across the top Hacker News stories.

Every `ctx.task(...)` is a durable checkpoint. If the worker crashes after some
stories have been summarized, the runtime re-runs only the missing ones — model
calls included.
"""
from __future__ import annotations

import asyncio

from agnt5 import WorkflowContext, workflow

from agnt5_quickstart.functions import (
    assemble_digest,
    fetch_story,
    fetch_top_ids,
    summarize,
)


@workflow
async def digest(ctx: WorkflowContext, limit: int = 5) -> dict:
    """Fetch the top `limit` HN stories, summarize them in parallel, assemble.

    Trace shape:
        digest (workflow)
        ├─ fetch_top_ids
        ├─ fetch_story (×N parallel)
        ├─ summarize   (×N parallel)
        └─ assemble_digest
    """
    ctx.logger.info("Starting digest for top %d stories", limit)

    # 1. One checkpoint: pull the IDs.
    ids = await ctx.task(fetch_top_ids, limit=limit)

    # 2. Fan out: each fetch_story is its own checkpoint. asyncio.gather runs
    #    them concurrently. A worker restart resumes from whichever of these
    #    had completed.
    stories = await asyncio.gather(*[
        ctx.task(fetch_story, story_id=i) for i in ids
    ])

    # 3. Fan out again on summarization. The Agent inside each `summarize`
    #    call passes `context=ctx`, so model calls also checkpoint.
    summaries = await asyncio.gather(*[
        ctx.task(summarize, story=s) for s in stories
    ])

    # 4. Combine. One last checkpoint, then return.
    return await ctx.task(assemble_digest, summaries=summaries)

File: typescript/quickstart/src/workflows.ts
/**
 * The `digest` workflow — fan out across the top Hacker News stories.
 *
 * Every step call is a durable checkpoint. If the worker crashes after some
 * stories have been summarized, the runtime re-runs only the missing ones —
 * model calls included.
 *
 *     digest (workflow)
 *     ├─ fetch_top_ids
 *     ├─ fetch_story  (×N parallel)
 *     ├─ summarize    (×N parallel)
 *     └─ assemble_digest
 */

import { workflow } from '@agnt5/sdk';
import type { Context } from '@agnt5/sdk';

import {
  assembleDigest,
  fetchStory,
  fetchTopIds,
  summarize,
} from './functions.js';

export const digest = workflow(
  'digest',
  async (ctx: Context, input: { limit?: number } = {}) => {
    const limit = input.limit ?? 5;
    ctx.logger.info(`Starting digest for top ${limit} stories`);

    // 1. Pull the IDs — one checkpoint.
    const ids = await fetchTopIds(ctx, { limit });

    // 2. Fan out: each fetchStory is its own checkpoint. Promise.all runs them
    //    concurrently. A worker restart resumes from whichever had completed.
    const stories = await Promise.all(
      ids.map((storyId) => fetchStory(ctx, { storyId })),
    );

    // 3. Fan out again on summarization. The Agent inside each summarize call
    //    threads `ctx`, so model calls also checkpoint.
    const summaries = await Promise.all(
      stories.map((story) => summarize(ctx, { story })),
    );

    // 4. Combine. One last checkpoint, then return.
    return assembleDigest(ctx, { summaries });
  },
);
